File: training/go/domain/model/creature.go
package model

import (
	"de.sots/cellarsandcentaurs/application"
)

var _ Fightable = (*Creature)(nil)

// A creature that roams the cellar. Centaurs, beasts and dragons all share hit points, armor class and speeds.
type Creature struct {
	id         CreatureId
	kind       CreatureType
	armorClass ArmorClass
	speeds     map[SpeedType]Speed
	hitPoints  HitPoints
}

func NewCreature(id CreatureId) *Creature {
	return NewCreatureOfType(id, application.StandardCreatureType)
}

func NewCreatureOfType(id CreatureId, kind CreatureType) *Creature {
	return &Creature{id: id, kind: kind, speeds: make(map[SpeedType]Speed)}
}

func (c *Creature) Id() CreatureId {
	return c.id
}

func (c *Creature) Type() CreatureType {
	return c.kind
}

func (c *Creature) SetType(kind CreatureType) {
	c.kind = kind
}

func (c *Creature) ArmorClass() ArmorClass {
	return c.armorClass
}

func (c *Creature) SetArmorClass(armorClass ArmorClass) {
	c.armorClass = armorClass
}

func (c *Creature) Speeds() map[SpeedType]Speed {
	return c.speeds
}

func (c *Creature) SetSpeeds(speeds map[SpeedType]Speed) {
	c.speeds = speeds
}

func (c *Creature) HitPoints() HitPoints {
	return c.hitPoints
}

func (c *Creature) SetHitPoints(hitPoints HitPoints) {
	c.hitPoints = hitPoints
}

func (c *Creature) TakeDamage(damage int) {
	c.hitPoints = c.hitPoints.TakeDamage(damage)
}

func (c *Creature) RollInitiative() DiceRoll {
	return rollD20()
}

File: training/php/src/Domain/Model/Creature.php
<?php

declare(strict_types=1);

namespace De\Sots\CellarsAndCentaurs\Domain\Model;

use De\Sots\CellarsAndCentaurs\Application\CreatureFacade;

/**
 * A creature that roams the cellar. Centaurs, beasts and dragons all share hit points, armor class and speeds.
 */
class Creature implements Fightable
{
    use HasSpeeds;

    private ?ArmorClass $armorClass = null;
    private ?HitPoints $hitPoints = null;

    public function __construct(
        private CreatureId $id,
        private CreatureType $type = CreatureFacade::STANDARD_CREATURE_TYPE
    ) {
    }

    public function getId(): CreatureId
    {
        return $this->id;
    }

    public function getType(): CreatureType
    {
        return $this->type;
    }

    public function setType(CreatureType $type): void
    {
        $this->type = $type;
    }

    public function getArmorClass(): ?ArmorClass
    {
        return $this->armorClass;
    }

    public function setArmorClass(ArmorClass $armorClass): void
    {
        $this->armorClass = $armorClass;
    }

    public function getHitPoints(): ?HitPoints
    {
        return $this->hitPoints;
    }

    public function setHitPoints(HitPoints $hitPoints): void
    {
        $this->hitPoints = $hitPoints;
    }

    public function takeDamage(int $damage): void
    {
        $this->hitPoints = $this->hitPoints?->damage($damage);
    }

    public function rollInitiative(): int
    {
        return rollD20()->getTotal();
    }
}
